Change the default gradient method logic to prefer devices, and to allow backprop on child devices

josh146 opened the pull request:
**Context:**

When using simulators, backpropagation scales significantly better for computing gradients than the parameter-shift rule as the number of parameters increases:

![image](https://user-images.githubusercontent.com/2959003/104911144-5ab25c80-59c5-11eb-9fef-34b89e381a78.png)

However, to use backprop, you need to explicitly load `default.qubit.tf` or `default.qubit.autograd`, which can cause the ability to perform backprop be a bit 'hidden'.

**Description of the Change:**

The logic for choosing the 'best' differentiation method has been altered to improve performance.

- If the device provides its own gradient, this is now the preferred differentiation method.

- If a device provides *child devices* that natively support classical backpropagation, this is now preferred over the parameter-shift rule.

  Devices define child devices via their `capabilities()` dictionary. For example, `default.qubit` supports child devices for TensorFlow, Autograd, and Jax:

  ```python
  {
    "passthru_devices": {
        "tf": "default.qubit.tf",
        "autograd": "default.qubit.autograd",
        "jax": "default.qubit.jax"
    },
  }
  ```

As a result of this change, if the QNode `diff_method` is not explicitly provided, it is possible that the QNode will run on a *child device* of the device that was specifically provided:

```python
dev = qml.device("default.qubit", wires=2)
qml.QNode(dev) # will default to backprop on default.qubit.autograd
qml.QNode(dev, interface="tf") # will default to backprop on default.qubit.tf
qml.QNode(dev, interface="jax") # will default to backprop on default.qubit.jax
```

**Benefits:**

* Users will see a speed up during optimization tasks, as backprop will now be the default rather than the parameter-shift if using `default.qubit`.

**Possible Drawbacks:**

* For circuits executed on the CPU, there can be some overhead on the forward pass when using TensorFlow, since intermediate values of the computation is being stored.

* The QNode is using a different device to that provided by the user (`dev`). As a result, calling `dev.state()` after executing the QNode will not provide expected results. However, you can still call `qnode.device.state`. Since we aim to deprecate the device methods from being user facing, I'm not too worried about this change.

* Backprop is only supported if `analytic=True`. If this is not the case, the logic simply fallsback to parameter-shift.

**Related GitHub Issues:** n/a

josh146 commented:
It looks like there are quite a few tests failing, because they don't specify `diff_method=` when creating the QNode, and just assume a parameter-shift QNode is being created.

These should be easily fixed by adding `diff_method="parameter-shift"` and making the tests explicit.

josh146 commented:
Note: this PR is worth benchmarking, to ensure that it provides a better experience for users using the default settings, e.g.,

```python
dev = qml.device("default.qubit", wires=2)
@qml.qnode(dev)
```

josh146 commented on pennylane/tape/qnode.py:
```diff
@@ -140,13 +140,16 @@ def __init__(self, func, device, interface="autograd", diff_method="best", **dif
+        self.diff_method = diff_method
 
-        self._tape, self.interface, self.diff_method = self.get_tape(device, interface, diff_method)
+        self._tape, self.interface, diff_method, self.device = self.get_tape(
+            device, interface, diff_method
+        )
```
Since it is now possible for the type of tape to be able to change the device :)

josh146 commented on pennylane/devices/default_qubit_autograd.py:
```diff
@@ -92,6 +92,7 @@ class DefaultQubitAutograd(DefaultQubit):
         "CRX": autograd_ops.CRX,
         "CRY": autograd_ops.CRY,
         "CRZ": autograd_ops.CRZ,
+        "CRot": autograd_ops.CRot,
```
Noticed this was missing due to tests suddenly failing in the new backprop default!

josh146 commented:
[ch4090] [ch4082]

trbromley commented on pennylane/tape/qnode.py:
```diff
@@ -249,21 +252,39 @@ def _validate_backprop_method(device, interface):
+            # device is analytic and has child devices that support backpropagation natively
+
+            if interface in backprop_devices:
+                # TODO: need a better way of passing existing device init options
+                # to a new device?
```
Good point :thinking:

I guess one option could be to define a `to_tf` interface in `DefaultQubit` that converts it to its TF version? This would also solve the problem of the user not being able to access the newly defined device.

josh146 replied:
That's a good idea... but I'm not sure how to implement it 🤔 

Changing the device requires changing the device class. We could do
```python
self.__class__ = passthru_device
```

but this is probably quite unsafe, especially if the `__init__` methods differ between the two.

trbromley replied:
Right! So is this ok to do for tapes (e.g., in `TFInterface.apply()`) because the `__init__` methods are the same?

josh146 replied:
yeah, `TFInterface`, `AutogradInterface` were designed from scratch with this capability in mind (they are mixin classes). Whereas we never designed the devices to be swappable 🤔

josh146 replied:
Some details: https://stackoverflow.com/a/13280789/10958457

trbromley commented on pennylane/tape/qnode.py:
```diff
@@ -140,13 +140,16 @@ def __init__(self, func, device, interface="autograd", diff_method="best", **dif
+        )
         self.diff_options = diff_options or {}
-        self.diff_options["method"] = self.diff_method
+        self.diff_options["method"] = diff_method
```
What's the difference between this one and `self.diff_method` :thinking:

josh146 replied:
See https://github.com/PennyLaneAI/pennylane/pull/1008/files#r559902787
```suggestion
        # Store the differentiation method to be passed to JacobianTape.jacobian().
        # Note that the tape accepts a different set of allowed methods than the QNode:
        #     best, analytic, numeric, device
        self.diff_options["method"] = diff_method
```

trbromley commented on pennylane/tape/qnode.py:
```diff
@@ -546,6 +567,10 @@ def to_tf(self, dtype=None):
             from pennylane.tape.interfaces.tf import TFInterface
 
             self.interface = "tf"
+            self._tape, self.interface, diff_method, self.device = self.get_tape(
```
Maybe worth adding a comment explaining this bit? Hasn't `get_tape` already been called upon instantiation?

josh146 replied:
Good point! The reason I added it here was to support the `qnn` module, where we make use of the fact that a QNode with an existing interface can be converted to another interface. Perhaps the following is less redundant?

```python
if self.interface != "tf":
    # Interface is changing, need to re-validate the tape class.
    self._tape, self.interface, diff_method, self.device = self.get_tape(
        self.original_device, self.interface, self.diff_method
    )
self.interface = "tf"
```

josh146 replied:
Have made this change :)

trbromley commented on pennylane/tape/qnode.py:
```diff
@@ -140,13 +140,16 @@ def __init__(self, func, device, interface="autograd", diff_method="best", **dif
 
         self.func = func
-        self.device = device
+        self.original_device = device
```
```suggestion
        self._original_device = device
```
Is this one for users to access?

trbromley reviewed (commented):
Thanks @josh146! Looks great, I had a few questions but would be happy to approve!

co9olguy commented:
> It looks like there are quite a few tests failing, because they don't specify `diff_method=` when creating the QNode, and just assume a parameter-shift QNode is being created.
> 
> These should be easily fixed by adding `diff_method="parameter-shift"` and making the tests explicit.

Isn't this also an indication that there are some "rough edges" in the codebase itself (not just the test suite)? Shouldn't all standard use-cases work without caring about the specific diff method?

Thenerdstation commented:
FYI: This change has awesome speedups for autograd, but seems to slow down the circuit evaluation a lot.

```
$ asv compare master default-gradient

All benchmarks:

       before           after         ratio
     [599e3186]       [b672829e]
     <master>         <default-gradient>
+      6.19±0.3ms       13.1±0.1ms     2.12  asv.core_suite.CircuitEvaluation.time_circuit(10, 3)
+      11.2±0.5ms       25.1±0.4ms     2.24  asv.core_suite.CircuitEvaluation.time_circuit(10, 6)
+      15.8±0.6ms       36.9±0.7ms     2.33  asv.core_suite.CircuitEvaluation.time_circuit(10, 9)
+     1.88±0.07ms       3.83±0.1ms     2.04  asv.core_suite.CircuitEvaluation.time_circuit(2, 3)
+      2.68±0.1ms       5.98±0.2ms     2.23  asv.core_suite.CircuitEvaluation.time_circuit(2, 6)
+      3.51±0.1ms       7.71±0.3ms     2.20  asv.core_suite.CircuitEvaluation.time_circuit(2, 9)
+      3.18±0.1ms       7.11±0.3ms     2.24  asv.core_suite.CircuitEvaluation.time_circuit(5, 3)
+      5.23±0.2ms      12.3±0.04ms     2.35  asv.core_suite.CircuitEvaluation.time_circuit(5, 6)
+      7.37±0.3ms       17.3±0.4ms     2.35  asv.core_suite.CircuitEvaluation.time_circuit(5, 9)
-      7.82±0.4ms       4.93±0.2ms     0.63  asv.core_suite.GradientComputation.time_gradient_autograd(2, 3)
-      21.6±0.7ms       8.09±0.3ms     0.37  asv.core_suite.GradientComputation.time_gradient_autograd(2, 6)
-        40.3±1ms       11.3±0.4ms     0.28  asv.core_suite.GradientComputation.time_gradient_autograd(5, 3)
-         142±2ms       19.4±0.8ms     0.14  asv.core_suite.GradientComputation.time_gradient_autograd(5, 6)
+      10.8±0.1ms       16.3±0.1ms     1.51  asv.core_suite.GradientComputation.time_gradient_tf(2, 3)
       27.0±0.3ms       29.4±0.3ms     1.09  asv.core_suite.GradientComputation.time_gradient_tf(2, 6)
       46.2±0.2ms         47.1±3ms     1.02  asv.core_suite.GradientComputation.time_gradient_tf(5, 3)
-         154±2ms       88.8±0.7ms     0.58  asv.core_suite.GradientComputation.time_gradient_tf(5, 6)
       8.01±0.3ms       8.11±0.2ms     1.01  asv.core_suite.GradientComputation.time_gradient_torch(2, 3)
       22.2±0.9ms       21.3±0.2ms     0.96  asv.core_suite.GradientComputation.time_gradient_torch(2, 6)
       39.3±0.6ms       39.4±0.1ms     1.00  asv.core_suite.GradientComputation.time_gradient_torch(5, 3)
        144±0.6ms          140±2ms     0.97  asv.core_suite.GradientComputation.time_gradient_torch(5, 6)

```

thisac commented on pennylane/tape/qnode.py:
```diff
@@ -249,21 +252,39 @@ def _validate_backprop_method(device, interface):
+            if interface in backprop_devices:
+                # TODO: need a better way of passing existing device init options
+                # to a new device?
+                device = qml.device(backprop_devices[interface], wires=device.wires, analytic=True)
```
(minor / idea) Would it perhaps be better to simply change `self.device` here directly instead of returning the device, storing the current device as default at init (making a comment that tapes may change devices; thus there's also an `original_device`)? Not sure if that's actually an improvement though.

```suggestion
                self.device = qml.device(backprop_devices[interface], wires=device.wires, analytic=True)

```

josh146 replied:
note that these validation methods are all 'pure' static methods - they do not access or change the state :) This is super useful for testing the logic in the unit tests

thisac replied:
I see. Makes sense to keep this then. :slightly_smiling_face:

thisac commented on tests/collections/test_collections.py:
```diff
@@ -59,7 +59,7 @@ def test_mapping_over_observables(self):
         template = lambda x, wires: qml.RX(x, wires=0)
 
-        qc = qml.map(template, obs_list, dev)
+        qc = qml.map(template, obs_list, dev, diff_method="parameter-shift")
```
Why is the parameter shift needed here to get the tests to pass? :thinking:

josh146 replied:
This is because of the assertion:

```python
assert qc[0].device is qc[1].device
```

🤔

thisac replied:
Oh, true, because the device is changed midway when using backprop. This seems a bit weird, and confusing to be honest. It has more to do with the way a device/child-device is chosen rather than that the parameter-shift method needs to be used. I feel like these tests perhaps shouldn't check that the device stays the same any more, since this i obviously not the case, rather than locking it to a specific diff method. :thinking:

josh146 replied:
Good call, maybe I will just remove the device assertions.

josh146 replied:
Have removed these :)

thisac commented on tests/tape/tapes/test_qnode.py:
```diff
@@ -52,16 +52,17 @@ def test_validate_device_method(self, monkeypatch):
         """Test that the method for validating the backprop diff method
         tape raises an exception if the device does not support backprop."""
-        dev = qml.device("default.qubit", wires=1)
+        dev = qml.device("default.gaussian", wires=1)
```
:+1:

thisac reviewed (commented):
Looks nice @josh146. Also curious about the things that @trbromley commented on. Only, not sure exactly why the parameter shift diff method needs to be set for some tests.

Also, noticed that the docstrings for `get_tape` , and some of the following methods, need to be updated. The `Returns:` should include the returned device and the priority order should be switched for backprop and device in `get_tape`.

josh146 commented:
> Isn't this also an indication that there are some "rough edges" in the codebase itself (not just the test suite)? Shouldn't all standard use-cases work without caring about the specific diff method?

@co9olguy: yes but also no!

* A majority of the cases (~75), the test was explicitly assuming parameter-shift without passing `diff_method="parameter-shift"`. That is, the test was explicitly mocking out a low level parameter-shift, or it was assuming an intermediate interface performing tensor unwrapping, or it was actually testing the parameter-shift gradient.

  In these cases, the tests should have been specifying `diff_method="parameter-shift"` anyway, so I added it in. This makes these tests more resilient to future changes of the default.

* There were ~30 tests where this was not the case. These tests were high-level integration tests that should have worked independently of the differentiation method. In these cases, I tracked down the bugs in the source code and fixed them, so that the tests passed without needing to specify `diff_method`.

Considering we have almost 6000 test cases, the fact that only 100 failed after changing the default gradient method was very relieving to me 🙂 I thought more would fail.

> FYI: This change has awesome speedups for autograd, but seems to slow down the circuit evaluation a lot.

@Thenerdstation: @mariaschuld noticed the same thing. Here are some more benchmarks (master on the left):

<div class="align-left">
<img src="https://user-images.githubusercontent.com/2959003/104986206-87aa5200-5a4d-11eb-9ba7-8fa8df774789.png" width=400px>
<img src="https://user-images.githubusercontent.com/2959003/104986214-8a0cac00-5a4d-11eb-9176-0cef209e8757.png" width=400px>
</div>

A couple of notes:

* The forward pass evaluation is slower
* Computing gradients with Autograd is waaay faster!
* Computing gradients with TF is slower

Just from intuition, I'm imagining that the slower forward pass is simply due to the overhead of storing intermediate statevectors in memory. I also imagine we have something like the following:

![image](https://user-images.githubusercontent.com/2959003/104986478-0a331180-5a4e-11eb-8d10-8cd1414c6c67.png)

That is, the forward pass overhead dominates for fewer parameters, but the parameter-shift scaling dominates for more parameters.

Question: Is this a better default?

* If the user is simply doing forward passes, then **no**. However, if they do not care about gradients, they could also specify `interface=None`, which removes all gradient overhead from the computation (this should even be faster that the parameter-shift current default).

* If the user is doing optimization, then this default should be faster.

josh146 commented on .github/CHANGELOG.md:
````diff
@@ -158,6 +158,40 @@
       return qml.expval(qml.PauliZ(0))
   ```
 
+* The logic for choosing the 'best' differentiation method has been altered
````
```suggestion
* In tape-mode, the logic for choosing the 'best' differentiation method has been altered
```
:)

josh146 commented on .github/CHANGELOG.md:
````diff
@@ -158,6 +158,40 @@
+
+    ```python
+    {
+      "passthru_devices": {
````
I like `passthru_devices` more, but we never use 'passthru' in user-facing documentation, so I wasn't sure whether to use it 🤔

josh146 replied:
have updated this to remove the word 'child'!

josh146 commented on pennylane/tape/qnode.py:
```diff
@@ -140,13 +140,16 @@ def __init__(self, func, device, interface="autograd", diff_method="best", **dif
+        self.original_device = device
         self.qtape = None
         self.qfunc_output = None
+        self.diff_method = diff_method
```
Ah, this is a subtlety that confused me as well. The issue is that the `diff_method` argument for the QNode, and the `diff_method` argument for the `Tape.jacobian()`:

* QNode: accepts `best`, `parameter-shift`, `finite-diff`, `device`, `backprop`
* Tape: accepts `best`, `numeric`, `analytic`, `numeric` 

🙁 

So we need to store both the QNode argument _and_ the `tape.jacobian` argument for further use. It's a bit annoying.

Here, `self.diff_method` stores the QNode argument (since I think that makes more sense), whereas `self.diff_options` stores the arguments given to `Tape.jacobian`.

```suggestion
        # store the user-specified differentiation method
        self.diff_method = diff_method
```

josh146 commented on pennylane/tape/qnode.py:
```diff
@@ -140,13 +140,16 @@ def __init__(self, func, device, interface="autograd", diff_method="best", **dif
+        self._tape, self.interface, diff_method, self.device = self.get_tape(
+            device, interface, diff_method
+        )
         self.diff_options = diff_options or {}
```
```suggestion
        # The arguments to be passed to JacobianTape.jacobian
        self.diff_options = diff_options or {}
```

josh146 commented on pennylane/tape/qnode.py:
```diff
@@ -546,6 +567,10 @@ def to_tf(self, dtype=None):
 
             self.interface = "tf"
+            self._tape, self.interface, diff_method, self.device = self.get_tape(
+                self.original_device, self.interface, self.diff_method
```
```suggestion
                self._original_device, self.interface, self.diff_method
```

josh146 commented on pennylane/tape/qnode.py:
```diff
@@ -577,6 +602,10 @@ def to_torch(self, dtype=None):
 
             self.interface = "torch"
+            self._tape, self.interface, diff_method, self.device = self.get_tape(
+                self.original_device, self.interface, self.diff_method
```
```suggestion
                self._original_device, self.interface, self.diff_method
```

trbromley commented on pennylane/tape/qnode.py:
```diff
@@ -140,20 +140,28 @@ def __init__(self, func, device, interface="autograd", diff_method="best", **dif
+        # The arguments to be passed to JacobianTape.jacobian
         self.diff_options = diff_options or {}
-        self.diff_options["method"] = self.diff_method
+        # Store the differentiation method to be passed to JacobianTape.jacobian().
+        # Note that the tape accepts a different set of allowed methods than the QNode:
+        #     best, analytic, numeric, device
```
No problem for this PR, but what is the reason for the difference in method names between the QNode and the Jacobian tape?

josh146 replied:
It's purely historical, they've always deviated 🙁 I wish it were not the case, and wouldn't mind changing it.

trbromley commented on pennylane/tape/qnode.py:
```diff
@@ -249,21 +257,39 @@ def _validate_backprop_method(device, interface):
         backprop_interface = device.capabilities().get("passthru_interface", None)
 
+        # determine if the device has any child devices that support backpropagation
+        backprop_devices = device.capabilities().get("passthru_devices", None)
```
Curious how this line wouldn't be covered by tests, when following ones are :thinking: Maybe it's a bug?
![image](https://user-images.githubusercontent.com/49409390/105103897-50549900-5a7f-11eb-9e5c-3abaa3a97fbf.png)

josh146 replied:
Yeah, this doesn't seem possible 🤔

trbromley reviewed (commented):
Thanks @josh146, I just want to have a quick go at testing this for some simple examples and then I can approve.

trbromley commented on pennylane/tape/qnode.py:
```diff
@@ -576,7 +611,16 @@ def to_torch(self, dtype=None):
-            self.interface = "torch"
+            if self.interface != "torch" and self.interface is not None:
+                # Since the interface is changing, need to re-validate the tape class.
+                self._tape, interface, diff_method, self.device = self.get_tape(
+                    self._original_device, "torch", self.diff_method
+                )
```
Ohh wait, I think I get these parts now! The idea is if we change the interface, it'll automatically check if it can change the device under the hood to ideally maintain backprop support?

Do we need to do this for torch?

josh146 replied:
Yep, because this allows us to go _backward_ 🙂 

E.g., a QNode running on `default.qubit.tf` and backprop will be able to be converted to a QNode running on default.qubit with parameter-shift!

Even better, if the user explicitly asked for backprop, the QNode remembers that preference (since we save it as `self.diff_method`), and calling `qnode.to_torch()` will raise an error since backprop is not supported. If `self.diff_method` was `best`, it will work with no issue.

trbromley reviewed (approved):
Thanks @josh146 :100:
